### PR TITLE
feat(datepicker): merge datepicker branch into master

### PR DESCRIPTION
**Pull request by mmalerba**
#675

### CODE REVIEW

**Comment by fxck** [comment]
@mmalerba don't you think https://github.com/angular/material2/issues/4358 should be fixed first? I just tested it again against the `angular/material2#datepicker` and it's still broken..

- use Czech locale using `setLocale('cs')`
- set default value to `2017-05-02T10:10:10.100Z'` (2nd of May)
- open datepicker and it opens on February
image

I'd say could imply there could be other locale related problems that tests aren't catching..

**Comment by mmalerba** [comment]
@jelbourn you may want to review these commits more closely: https://github.com/angular/material2/pull/4404/files/14b252336f126d82178a66224f614b6187f2dbc2..0697350b6803a66b0a044888d9f90f615f3d1b84 these are ones I made on top of the already-reviewed changes to fix tests and things.

File: src/lib/core/datetime/native-date-adapter.ts [hunk 1]
```diff
@@ -188,4 +188,15 @@ export class NativeDateAdapter extends DateAdapter<Date> {
   private _2digit(n: number) {
     return ('00' + n).slice(-2);
   }
+
+  /**
+   * Strip out unicode LTR and RTL characters. Edge and IE insert these into formatted dates while
+   * other browsers do not. We remove them to make output consistent and because they interfere with
+   * date parsing.
+   * @param s The string to strip direction characters from.
+   * @returns {string} The stripped string.
```
Comment (jelbourn) [inline src/lib/core/datetime/native-date-adapter.ts]: Omit return type from comment

File: src/lib/core/datetime/native-date-adapter.ts [hunk 2]
```diff
@@ -188,4 +188,15 @@ export class NativeDateAdapter extends DateAdapter<Date> {
...
+   */
+  private _stripDirectionCharacters(s: string) {
```
Comment (jelbourn) [inline src/lib/core/datetime/native-date-adapter.ts]: `_stripDirectionalityCharacters`? (to match our pending terminology)

**Comment by mmalerba** [comment]
Done

**Pull request merged**